File: app/src/main/java/fr/free/nrw/commons/campaigns/CampaignsPresenter.java
package fr.free.nrw.commons.campaigns;

import android.annotation.SuppressLint;

import java.text.ParseException;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import fr.free.nrw.commons.BasePresenter;
import fr.free.nrw.commons.mwapi.OkHttpJsonApiClient;
import fr.free.nrw.commons.utils.CommonsDateUtil;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;
import timber.log.Timber;

import static fr.free.nrw.commons.di.CommonsApplicationModule.IO_THREAD;
import static fr.free.nrw.commons.di.CommonsApplicationModule.MAIN_THREAD;

/**
 * The presenter for the campaigns view, fetches the campaigns from the api and informs the view on
 * success and error
 */
@Singleton
public class CampaignsPresenter implements BasePresenter<ICampaignsView> {
    private final OkHttpJsonApiClient okHttpJsonApiClient;
    private final Scheduler mainThreadScheduler;
    private final Scheduler ioScheduler;

    private ICampaignsView view;
    private Disposable disposable;
    private Campaign campaign;

    @Inject
    public CampaignsPresenter(OkHttpJsonApiClient okHttpJsonApiClient, @Named(IO_THREAD)Scheduler ioScheduler, @Named(MAIN_THREAD)Scheduler mainThreadScheduler) {
        this.okHttpJsonApiClient = okHttpJsonApiClient;
        this.mainThreadScheduler=mainThreadScheduler;
        this.ioScheduler=ioScheduler;
    }

    @Override
    public void onAttachView(ICampaignsView view) {
        this.view = view;
    }

    @Override public void onDetachView() {
        this.view = null;
        if (disposable != null) {
            disposable.dispose();
        }
    }

    /**
     * make the api call to fetch the campaigns
     */
    @SuppressLint("CheckResult")
    public void getCampaigns() {
        if (view != null && okHttpJsonApiClient != null) {
            //If we already have a campaign, lets not make another call
            if (this.campaign != null) {
                view.showCampaigns(campaign);
                return;
            }
            Single<CampaignResponseDTO> campaigns = okHttpJsonApiClient.getCampaigns();
            campaigns.observeOn(mainThreadScheduler)
                .subscribeOn(ioScheduler)
                .subscribeWith(new SingleObserver<CampaignResponseDTO>() {

                    @Override public void onSubscribe(Disposable d) {
                        disposable = d;
                    }

                    @Override public void onSuccess(CampaignResponseDTO campaignResponseDTO) {
                        List<Campaign> campaigns = campaignResponseDTO.getCampaigns();
                        if (campaigns == null || campaigns.isEmpty()) {
                            Timber.e("The campaigns list is empty");
                            view.showCampaigns(null);
                            return;
                        }
                        Collections.sort(campaigns, (campaign, t1) -> {
                            Date date1, date2;
                            try {

                                date1 = CommonsDateUtil.getIso8601DateFormatShort().parse(campaign.getStartDate());
                                date2 = CommonsDateUtil.getIso8601DateFormatShort().parse(t1.getStartDate());
                            } catch (ParseException e) {
                                e.printStackTrace();
                                return -1;
                            }
                            return date1.compareTo(date2);
                        });
                        Date campaignEndDate, campaignStartDate;
                        Date currentDate = new Date();
                        try {
                            for (Campaign aCampaign : campaigns) {
                                campaignEndDate = CommonsDateUtil.getIso8601DateFormatShort().parse(aCampaign.getEndDate());
                                campaignStartDate = CommonsDateUtil.getIso8601DateFormatShort().parse(aCampaign.getStartDate());
                                if (campaignEndDate.compareTo(currentDate) >= 0
                                    && campaignStartDate.compareTo(currentDate) <= 0) {
                                    campaign = aCampaign;
                                    break;
                                }
                            }
                        } catch (ParseException e) {
                            e.printStackTrace();
                        }
                        view.showCampaigns(campaign);
                    }

                    @Override public void onError(Throwable e) {
                        Timber.e(e, "could not fetch campaigns");
                    }
                });
        }
    }
}

File: app/src/main/java/fr/free/nrw/commons/campaigns/CampaignsPresenter.kt
package fr.free.nrw.commons.campaigns

import android.annotation.SuppressLint
import fr.free.nrw.commons.BasePresenter
import fr.free.nrw.commons.campaigns.models.Campaign
import fr.free.nrw.commons.di.CommonsApplicationModule.IO_THREAD
import fr.free.nrw.commons.di.CommonsApplicationModule.MAIN_THREAD
import fr.free.nrw.commons.mwapi.OkHttpJsonApiClient
import fr.free.nrw.commons.utils.CommonsDateUtil.getIso8601DateFormatShort
import io.reactivex.Scheduler
import io.reactivex.disposables.Disposable
import timber.log.Timber
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * The presenter for the campaigns view, fetches the campaigns from the api and informs the view on
 * success and error
 */
@Singleton
class CampaignsPresenter @Inject constructor(
    private val okHttpJsonApiClient: OkHttpJsonApiClient?,
    @param:Named(IO_THREAD) private val ioScheduler: Scheduler,
    @param:Named(MAIN_THREAD) private val mainThreadScheduler: Scheduler
) : BasePresenter<ICampaignsView> {
    private var view: ICampaignsView? = null
    private var disposable: Disposable? = null
    private var campaign: Campaign? = null

    override fun onAttachView(view: ICampaignsView) {
        this.view = view
    }

    override fun onDetachView() {
        view = null
        disposable?.dispose()
    }

    /**
     * make the api call to fetch the campaigns
     */
    @SuppressLint("CheckResult")
    fun getCampaigns() {
        if (view != null && okHttpJsonApiClient != null) {
            //If we already have a campaign, lets not make another call
            if (campaign != null) {
                view!!.showCampaigns(campaign)
                return
            }

            okHttpJsonApiClient.getCampaigns()
                .observeOn(mainThreadScheduler)
                .subscribeOn(ioScheduler)
                .doOnSubscribe { disposable = it }
                .subscribe({ campaignResponseDTO ->
                    val campaigns = campaignResponseDTO?.campaigns?.toMutableList()
                    if (campaigns.isNullOrEmpty()) {
                        Timber.e("The campaigns list is empty")
                        view!!.showCampaigns(null)
                    } else {
                        sortCampaignsByStartDate(campaigns)
                        campaign = findActiveCampaign(campaigns)
                        view!!.showCampaigns(campaign)
                    }
                }, {
                    Timber.e(it, "could not fetch campaigns")
                })
        }
    }

    private fun sortCampaignsByStartDate(campaigns: MutableList<Campaign>) {
        val dateFormat: SimpleDateFormat = getIso8601DateFormatShort()
        campaigns.sortWith(Comparator { campaign: Campaign, other: Campaign ->
            val date1: Date?
            val date2: Date?
            try {
                date1 = campaign.startDate?.let { dateFormat.parse(it) }
                date2 = other.startDate?.let { dateFormat.parse(it) }
            } catch (e: ParseException) {
                Timber.e(e)
                return@Comparator -1
            }
            if (date1 != null && date2 != null) date1.compareTo(date2) else -1
        })
    }

    private fun findActiveCampaign(campaigns: List<Campaign>) : Campaign? {
        val dateFormat: SimpleDateFormat = getIso8601DateFormatShort()
        val currentDate = Date()
        return try {
            campaigns.firstOrNull {
                val campaignStartDate = it.startDate?.let { s -> dateFormat.parse(s) }
                val campaignEndDate = it.endDate?.let { s -> dateFormat.parse(s) }
                campaignStartDate != null && campaignEndDate != null &&
                        campaignEndDate >= currentDate && campaignStartDate <= currentDate
            }
        } catch (e: ParseException) {
            Timber.e(e, "could not find active campaign")
            null
        }
    }
}
